kafka-client/src/main/java/dev/responsive/kafka/internal/metrics/CassandraMetricsFactory.java

```diff
@@ -0,0 +1,251 @@
+ * </p>
+ */
+public class CassandraMetricsFactory implements MetricsFactory {
+  static final String CASSANDRA_METRICS_GROUP = "cassandra-driver";
+  private final SessionMetricUpdater sessionMetricUpdater;
+
+  public CassandraMetricsFactory(final DriverContext driverContext) {
+    this.sessionMetricUpdater = fetchMetrics(driverContext)
+        .map(metrics -> (SessionMetricUpdater) new CassandraMetricUpdater(metrics))
+        .orElse(NoopSessionMetricUpdater.INSTANCE);
+  }
+
+  private static Optional<ResponsiveMetrics> fetchMetrics(DriverContext driverContext) {
+    // Note that there seems to be no alternative to accessing the metric registry
+    // than `InternalDriverContext`. Each of the included metric factory classes
+    // cast in order to get access.
+    if (driverContext instanceof InternalDriverContext) {
+      final InternalDriverContext internalDriverContext = (InternalDriverContext) driverContext;
+      final Object metricRegistry = internalDriverContext.getMetricRegistry();
+      if (metricRegistry instanceof ResponsiveMetrics) {
+        return Optional.of((ResponsiveMetrics) internalDriverContext.getMetricRegistry());
+      }
+    }
+    return Optional.empty();
+  }
+
+  @Override
+  public Optional<Metrics> getMetrics() {
+    return Optional.empty();
+  }
+
+  @Override
+  public SessionMetricUpdater getSessionUpdater() {
+    return sessionMetricUpdater;
+  }
+
+  @Override
+  public NodeMetricUpdater newNodeUpdater(final Node node) {
+    return NoopNodeMetricUpdater.INSTANCE;
+  }
+
+  private static class CassandraMetricUpdater implements SessionMetricUpdater {
+    private final ResponsiveMetrics metrics;
+
+    private CassandraMetricUpdater(final ResponsiveMetrics metrics) {
+      this.metrics = metrics;
+    }
+
+    @Override
+    public void incrementCounter(
+        final SessionMetric sessionMetric,
+        @Nullable final String profileName,
+        final long amount
+    ) {
+      getOrCreateSensor(sessionMetric).record(amount);
+    }
+
+    @Override
+    public void updateHistogram(
+        final SessionMetric sessionMetric,
+        @Nullable final String s,
+        final long value
+    ) {
+      getOrCreateSensor(sessionMetric).record(value);
+    }
+
+    @Override
+    public void markMeter(
+        final SessionMetric sessionMetric,
+        @Nullable final String profileName,
+        final long amount
+    ) {
+      getOrCreateSensor(sessionMetric).record(amount);
+    }
+
+    @Override
+    public void updateTimer(
+        final SessionMetric sessionMetric,
+        @Nullable final String profileName,
+        final long duration,
+        final TimeUnit timeUnit
+    ) {
+      getOrCreateSensor(sessionMetric)
+          .record(TimeUnit.MILLISECONDS.convert(duration, timeUnit));
+    }
+
+    @Override
+    public boolean isEnabled(
+        final SessionMetric metric,
+        @Nullable final String profileName
+    ) {
+      switch (asDefaultSessionMetric(metric)) {
```

Comment:
do you know if there are any general error/success metrics or latency metrics?

Reply:
I could not find any general success/error metrics. There are some more error metrics defined under `DefaultNodeMetric`. I was a little concerned about exposing node level metrics, but the `MetricsFactory` would allow us to return a session-level metric as a node metric. In other words, we could just ignore the node. That would give us these metrics:
```java
  ABORTED_REQUESTS("errors.request.aborted"),
  WRITE_TIMEOUTS("errors.request.write-timeouts"),
  READ_TIMEOUTS("errors.request.read-timeouts"),
  UNAVAILABLES("errors.request.unavailables"),
  OTHER_ERRORS("errors.request.others"),
  CONNECTION_INIT_ERRORS("errors.connection.init"),
  AUTHENTICATION_ERRORS("errors.connection.auth");
```
What do you think?

Reply:
Ugh that's pretty verbose... let's just keep it as you have in your PR for now and if we end up having any issues we can work them in later. Of those I think timeouts are the most interesting.

Reply:
Also not sure how I missed `cql-requests-cumulative-latency` but looks like that's already there.

Reply:
It would be to roll up all of the error metrics into a single `errors.total` metric. We could also expose these:
```
        # The total number of errors on this node that caused the RetryPolicy to trigger a retry
        # (exposed as a Counter).
        #
        # This is a sum of all the other retries.* metrics.
        // retries.total,
        # The total number of errors on this node that were ignored by the RetryPolicy (exposed as
        # a Counter).
        #
        # This is a sum of all the other ignores.* metrics.
        // ignores.total,
```
That would give us a pretty good overall picture about error and retry behavior.

Reply:
💯 that would be perfect

Reply:
Sounds good. I'll update the patch.